### DISCUSSION TITLE
[BUG] `ImportError: DLL load failed while importing pytransform3`

### DISCUSSION THREAD

**Discussion by pauloswear** [dashingsoft/pyarmor#1257] [Q&A]
**Platform:**
```
Windows x64
Python 3.10.5 x86
```

**Description:**
```
Trying to encrypt using BCC mode. The objective is encrypt to all systems (x64 and x86), that's the reason i am using python 3.10.5 x86
```

**How to reproduce:**
```
Install any x86 python version and try encrypt with this CLI
pyarmor gen  script.py --recursive --enable-bcc --mix-str --assert-call --assert-import 
```

Error:
`27592 MainProcess 2023-05-17 16:56:05,891
Traceback (most recent call last):
  File "C:\Users\Paulo\AppData\Local\Programs\Python\Python310-32\lib\site-packages\pyarmor\cli\__main__.py", line 706, in main
    main_entry(sys.argv[1:])
  File "C:\Users\Paulo\AppData\Local\Programs\Python\Python310-32\lib\site-packages\pyarmor\cli\__main__.py", line 694, in main_entry
    args.func(ctx, args)
  File "C:\Users\Paulo\AppData\Local\Programs\Python\Python310-32\lib\site-packages\pyarmor\cli\__main__.py", line 237, in cmd_gen
    builder.process(options)
  File "C:\Users\Paulo\AppData\Local\Programs\Python\Python310-32\lib\site-packages\pyarmor\cli\generate.py", line 147, in process
    Pytransform3.pre_build(self.ctx)
  File "C:\Users\Paulo\AppData\Local\Programs\Python\Python310-32\lib\site-packages\pyarmor\cli\core\__init__.py", line 79, in pre_build
    m = Pytransform3.init(ctx)
  File "C:\Users\Paulo\AppData\Local\Programs\Python\Python310-32\lib\site-packages\pyarmor\cli\core\__init__.py", line 57, in init
    Pytransform3._pytransform3 = m = _import_pytransform3()
  File "C:\Users\Paulo\AppData\Local\Programs\Python\Python310-32\lib\site-packages\pyarmor\cli\core\__init__.py", line 34, in _import_pytransform3
    return __import__(
ImportError: DLL load failed while importing pytransform3: The specified procedure could not be found.
`

**CMD ERROR**
```
INFO     Python 3.10.5
INFO     Pyarmor 8.2.0 (pro), &&&, *****
INFO     Platform windows.x86
INFO     search inputs ...
INFO     find script launcher.py
INFO     find 1 top resources
ERROR    DLL load failed while importing pytransform3: The specified procedure could not be found.
```

**Comment by jondy** [comment]
Please also install `pyarmor.core` for windows.x86

**Reply by pauloswear** [reply]
Sorry, but how can I do it? there's nothing in docs

**Reply by VeNoMouS** [reply]
```
pip install -U pyarmor.core
```
?

**Reply by pauloswear** [reply]
This does not exists in pypi

**Comment by jondy** [comment]
@pauloswear `pyarmor.cli.core` is platform-dependent, windows.x86_64 is different from windows.x86, please also install `pyarmor.cli.core` for Python x86. Suppose Python x86 at `c:/Python310-32/python`, then
```
c:/Python310-32/python -m pip install pyarmor.cli.core
```

@VeNoMouS sorry, I made a mistake, the package name is `pyarmor.cli.core`, not `pyarmor.core`.

**Reply by pauloswear** [reply]
@jondy it was already installed

```
C:\Users\Paulo>python -m pip install pyarmor.cli.core
Requirement already satisfied: pyarmor.cli.core in c:\users\paulo\appdata\local\programs\python\python310-32\lib\site-packages (3.2.0)
```

But I still can't register pyarmor or use pyarmor gen

```
INFO     Python 3.10.5
INFO     Pyarmor 8.2.0 (trial), 000000, non-profits
INFO     Platform windows.x86
INFO     register "E:\My Drive\pyarmor-regfile-**.zip"
INFO     update license token
ERROR    DLL load failed while importing pytransform3: The specified procedure could not be found.
```

**Reply by jondy** [reply]
This python310-32 is debug-version or not? Is it downloaded from python.org?

**Reply by pauloswear** [reply]
It's official version. It's downloaded from python.org

**Reply by jondy** [reply]
How about importing pytransform3 directly?
```
C:/Python310-32/python
>>> import pyarmor.cli.core.pytransform3
```

**Reply by jondy** [reply]
After search error message `The specified procedure could not be found.` in msdn, I think the problem is that some dependent dlls of `pytransform3.pyd` are missed, please check it by `dumpbin`

**Reply by pauloswear** [reply]
> Python 3.10.5 (tags/v3.10.5:f377153, Jun  6 2022, 15:58:59) [MSC v.1929 32 bit (Intel)] on win32
> Type "help", "copyright", "credits" or "license" for more information.
> >>> import pyarmor.cli.core.pytransform3
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
> ImportError: DLL load failed while importing pytransform3: The specified procedure could not be found.

**Reply by pauloswear** [reply]
In other machine, same error ocurrs using windows 11

> 
> C:\Users\User>python
> Python 3.10.5 (tags/v3.10.5:f377153, Jun  6 2022, 15:58:59) [MSC v.1929 32 bit (Intel)] on win32
> Type "help", "copyright", "credits" or "license" for more information.
> >>> import pyarmor.cli.core.pytransform3
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
> ImportError: DLL load failed while importing pytransform3: Não foi possível encontrar o procedimento especificado.
> >>>

**Reply by pauloswear** [reply]
3rd machine (azure virtual private server ws2016) got error too

**Comment by jondy** [comment]
Download `vc_redist.x86.exe` and install it may fix this issue
https://www.microsoft.com/en-us/download/details.aspx?id=48145

**Reply by pauloswear** [reply]
not fixed

**Comment by pauloswear** [comment]
It seems that's some incompatibility with 3.10.x or HIGHER versions


Works:

```
Python 3.9.0 (tags/v3.9.0:9cf6752, Oct  5 2020, 15:23:07) [MSC v.1927 32 bit (Intel)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyarmor.cli.core.pytransform3
```

Not work:

3.11.x
```
Python 3.11.0 (main, Oct 24 2022, 18:13:38) [MSC v.1933 32 bit (Intel)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyarmor.cli.core.pytransform3
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: DLL load failed while importing pytransform3: The specified procedure could not be found.
```

3.10.x
```
Python 3.10.5 (tags/v3.10.5:f377153, Jun 6 2022, 15:58:59) [MSC v.1929 32 bit (Intel)] on win32
Type "help", "copyright", "credits" or "license" for more information.

import pyarmor.cli.core.pytransform3
Traceback (most recent call last):
File "", line 1, in
ImportError: DLL load failed while importing pytransform3: The specified procedure could not be found.
```

Can you check?

**Comment by jondy** [comment]
I have no win10 and Python 3.10+ doesn't work in win7.

Could you check the dependent libraries?
```
dumpbin /DEPENDENTS /path/to/python3.10/site-packages/pyarmor/cli/core/pytransform3.pyd
```

**Reply by pauloswear** [reply]
You can use this simple mod:
https://github.com/adang1345/PythonWin7

**Comment by jondy** [comment]
@pauloswear I found something different from python3.9, I think it need to rebuild python3.10 and python3.11 to fix this issue.

**Reply by pauloswear** [reply]
Rebuild python or pyarmor?

**Comment by jondy** [comment]
@pauloswear I rebuild them and upload to
http://dashingsoft.com/downloads/temp/cp3.10/pytransform3.pyd
http://dashingsoft.com/downloads/temp/cp3.10/pyarmor_runtime.pyd
http://dashingsoft.com/downloads/temp/cp3.11/pytransform3.pyd
http://dashingsoft.com/downloads/temp/cp3.11/pyarmor_runtime.pyd

Could you download them and replace the old ones and verify it works or not?

**Reply by pauloswear** [reply]
@jondy It solved, thanks. Will open different topic for other questions
